packages/core/base/src/signer.ts

```diff
@@ -1,42 +1,78 @@
-import type { Connection, Transaction, TransactionSignature } from '@solana/web3.js';
-import type { SendTransactionOptions, WalletAdapter } from './adapter.js';
+import type { Connection, TransactionSignature } from '@solana/web3.js';
+import type { SendTransactionOptions, WalletAdapter, WalletAdapterProps } from './adapter.js';
 import { BaseWalletAdapter } from './adapter.js';
 import { WalletSendTransactionError, WalletSignTransactionError } from './errors.js';
+import type { TransactionOrVersionedTransaction } from './types.js';
 
-export interface SignerWalletAdapterProps {
-    signTransaction(transaction: Transaction): Promise<Transaction>;
-    signAllTransactions(transaction: Transaction[]): Promise<Transaction[]>;
+export interface SignerWalletAdapterProps<Name extends string = string> extends WalletAdapterProps<Name> {
+    signTransaction<T extends TransactionOrVersionedTransaction<this['supportedTransactionVersions']>>(
+        transaction: T
+    ): Promise<T>;
+    signAllTransactions<T extends TransactionOrVersionedTransaction<this['supportedTransactionVersions']>>(
+        transactions: T[]
+    ): Promise<T[]>;
```

Comment:
Okay, I think these types are very reasonable now. By inferring `SupportedTransactionVersions` from the actual implementer, and making the methods return the type they were called with, existing code that expects `signTransaction(transaction: Transaction): Promise<Transaction>` will just work.